遇到 checkbox 时报错 panic: runtime error: invalid memory address or nil pointer

**问题描述**

发现遇到 checkbox（即下图中的 test2 一行）就无法正常导出，报错 `panic: runtime error: invalid memory address or nil pointer dereference`。已阅读过 #2 和 #3 ，未能解决。有没有可能是因为 markdown 的 checkbox 一般是 `- [ ]` 而飞书的是 `[]`？

![image](https://user-images.githubusercontent.com/45259695/216973659-7bc688b0-f523-4617-bf05-5a96a217fcea.png)

**To Reproduce**

[https://q6hxyvlrdh.feishu.cn/docx/KmnQddpS3oVUmkxqsGDcvI2Nnvc](https://q6hxyvlrdh.feishu.cn/docx/KmnQddpS3oVUmkxqsGDcvI2Nnvc)，即上述截图中的文档

**期望的 markdown 格式**

---

# test

test1

- [ ] test2

即

```md
# test

test1

- [ ] test2
```

---

**报错的截图**

```sh
Captured document token: KmnQddpS3oVUmkxqsGDcvI2Nnvc
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0xac81b0]

goroutine 1 [running]:
github.com/Wsine/feishu2md/core.(*Parser).ParseDocxBlock(0xc00029fb90, 0xc0003ae420, 0xc00029fb90)
        /github/workspace/core/parser.go:369 +0x9f0
github.com/Wsine/feishu2md/core.(*Parser).ParseDocxContent(0xc000006520, 0xe4ae20, {0xc0003950c8, 0x3, 0x1b})
        /github/workspace/core/parser.go:279 +0x14c
main.handleUrlArgument({0xc000010340, 0x3d}, 0x0)
        /github/workspace/main.go:97 +0x38b
main.main.func1(0xc0002b63c0)
        /github/workspace/main.go:151 +0xbd
github.com/urfave/cli/v2.(*App).RunContext(0xc000247ba0, {0xe4adb0, 0xc00000a070}, {0xc00002a3e0, 0x2, 0x2})
        /go/pkg/mod/github.com/urfave/cli/v2@v2.6.0/app.go:322 +0x7a8
github.com/urfave/cli/v2.(*App).Run(...)
        /go/pkg/mod/github.com/urfave/cli/v2@v2.6.0/app.go:224
main.main()
        /github/workspace/main.go:182 +0x394
```

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -19,6 +19,8 @@ jobs:
     steps:
       - name: Checkout the code
         uses: actions/checkout@v2
+      - name: Create version file
+        run: echo ${{ github.event.release.tag_name }} > VERSION
       - name: Parallel build
         uses: wangyoucao577/go-release-action@v1.30
         with:
@@ -42,6 +44,8 @@ jobs:
     steps:
       - name: Checkout the code
         uses: actions/checkout@v2
+      - name: Create version file
+        run: echo ${{ github.event.release.tag_name }} > VERSION
       - name: Parallel build
         uses: wangyoucao577/go-release-action@v1.30
         with:

--- a/.gitignore
+++ b/.gitignore
@@ -17,5 +17,6 @@ bin/
 feishu2md-amd64.zip
 .env
 static
+VERSION
 
 .idea

--- a/core/parser.go
+++ b/core/parser.go
@@ -382,6 +382,8 @@ func (p *Parser) ParseDocxBlock(b *lark.DocxBlock, blockMap *orderedmap.OrderedM
 		buf.WriteString(p.ParseDocxBlockTableCell(b.BlockID, blockMap))
 	case lark.DocxBlockTypeTable:
 		buf.WriteString(p.ParseDocxBlockTable(b.ParentID, b.Table, blockMap))
+  case lark.DocxBlockTypeQuoteContainer:
+		buf.WriteString(p.ParseDocxBlockQuoteContainer(b.BlockID, b.QuoteContainer, blockMap))
 	default:
 		return ""
 	}
@@ -507,4 +509,28 @@ func (p *Parser) ParseDocxBlockTable(documentId string, t *lark.DocxBlockTable, 
 	buf.WriteString(renderMarkdownTable(rows))
 	buf.WriteString("\n")
 	return buf.String()
+}
+
+func (p *Parser) ParseDocxBlockQuoteContainer(blockId string, q *lark.DocxBlocQuoteContainer, blockMap *orderedmap.OrderedMap) string {
+  contents := "> "
+	for _, key := range blockMap.Keys() {
+		value, ok := blockMap.Get(key)
+		if !ok {
+			continue
+		}
+		block := value.(*lark.DocxBlock)
+		if block.ParentID != blockId {
+			continue
+		}
+
+		content := p.ParseDocxBlock(block, blockMap)
+		if content == "" {
+			continue
+		}
+		contents += content
+		// remove quote container children block from map
+		blockMap.Delete(block.BlockID)
+	}
+	contents = strings.Join(strings.Fields(strings.ReplaceAll(strings.TrimSpace(strip.StripTags(contents)), "\n", "<br/>")), " ")
+	return contents
 }

--- a/core/parser.go
+++ b/core/parser.go
@@ -28,6 +28,76 @@ func NewParser(ctx context.Context) *Parser {
 // Parser utils
 // =============================================================
 
+var DocxCodeLang2MdStr = map[lark.DocxCodeLanguage]string{
+	lark.DocxCodeLanguagePlainText:    "",
+	lark.DocxCodeLanguageABAP:         "abap",
+	lark.DocxCodeLanguageAda:          "ada",
+	lark.DocxCodeLanguageApache:       "apache",
+	lark.DocxCodeLanguageApex:         "apex",
+	lark.DocxCodeLanguageAssembly:     "assembly",
+	lark.DocxCodeLanguageBash:         "bash",
+	lark.DocxCodeLanguageCSharp:       "csharp",
+	lark.DocxCodeLanguageCPlusPlus:    "cpp",
+	lark.DocxCodeLanguageC:            "c",
+	lark.DocxCodeLanguageCOBOL:        "cobol",
+	lark.DocxCodeLanguageCSS:          "css",
+	lark.DocxCodeLanguageCoffeeScript: "coffeescript",
+	lark.DocxCodeLanguageD:            "d",
+	lark.DocxCodeLanguageDart:         "dart",
+	lark.DocxCodeLanguageDelphi:       "delphi",
+	lark.DocxCodeLanguageDjango:       "django",
+	lark.DocxCodeLanguageDockerfile:   "dockerfile",
+	lark.DocxCodeLanguageErlang:       "erlang",
+	lark.DocxCodeLanguageFortran:      "fortran",
+	lark.DocxCodeLanguageFoxPro:       "foxpro",
+	lark.DocxCodeLanguageGo:           "go",
+	lark.DocxCodeLanguageGroovy:       "groovy",
+	lark.DocxCodeLanguageHTML:         "html",
+	lark.DocxCodeLanguageHTMLBars:     "htmlbars",
+	lark.DocxCodeLanguageHTTP:         "http",
+	lark.DocxCodeLanguageHaskell:      "haskell",
+	lark.DocxCodeLanguageJSON:         "json",
+	lark.DocxCodeLanguageJava:         "java",
+	lark.DocxCodeLanguageJavaScript:   "javascript",
+	lark.DocxCodeLanguageJulia:        "julia",
+	lark.DocxCodeLanguageKotlin:       "kotlin",
+	lark.DocxCodeLanguageLateX:        "latex",
+	lark.DocxCodeLanguageLisp:         "lisp",
+	lark.DocxCodeLanguageLogo:         "logo",
+	lark.DocxCodeLanguageLua:          "lua",
+	lark.DocxCodeLanguageMATLAB:       "matlab",
+	lark.DocxCodeLanguageMakefile:     "makefile",
+	lark.DocxCodeLanguageMarkdown:     "markdown",
+	lark.DocxCodeLanguageNginx:        "nginx",
+	lark.DocxCodeLanguageObjective:    "objectivec",
+	lark.DocxCodeLanguageOpenEdgeABL:  "openedge-abl",
+	lark.DocxCodeLanguagePHP:          "php",
+	lark.DocxCodeLanguagePerl:         "perl",
+	lark.DocxCodeLanguagePostScript:   "postscript",
+	lark.DocxCodeLanguagePower:        "powershell",
+	lark.DocxCodeLanguageProlog:       "prolog",
+	lark.DocxCodeLanguageProtoBuf:     "protobuf",
+	lark.DocxCodeLanguagePython:       "python",
+	lark.DocxCodeLanguageR:            "r",
+	lark.DocxCodeLanguageRPG:          "rpg",
+	lark.DocxCodeLanguageRuby:         "ruby",
+	lark.DocxCodeLanguageRust:         "rust",
+	lark.DocxCodeLanguageSAS:          "sas",
+	lark.DocxCodeLanguageSCSS:         "scss",
+	lark.DocxCodeLanguageSQL:          "sql",
+	lark.DocxCodeLanguageScala:        "scala",
+	lark.DocxCodeLanguageScheme:       "scheme",
+	lark.DocxCodeLanguageScratch:      "scratch",
+	lark.DocxCodeLanguageShell:        "shell",
+	lark.DocxCodeLanguageSwift:        "swift",
+	lark.DocxCodeLanguageThrift:       "thrift",
+	lark.DocxCodeLanguageTypeScript:   "typescript",
+	lark.DocxCodeLanguageVBScript:     "vbscript",
+	lark.DocxCodeLanguageVisual:       "vbnet",
+	lark.DocxCodeLanguageXML:          "xml",
+	lark.DocxCodeLanguageYAML:         "yaml",
+}
+
 func renderMarkdownTable(data [][]string) string {
 	builder := &strings.Builder{}
 	table := tablewriter.NewWriter(builder)
@@ -359,7 +429,7 @@ func (p *Parser) ParseDocxBlock(b *lark.DocxBlock, blockMap *orderedmap.OrderedM
 		buf.WriteString("1. ")
 		buf.WriteString(p.ParseDocxBlockText(b.Ordered))
 	case lark.DocxBlockTypeCode:
-		buf.WriteString("```\n")
+		buf.WriteString("```" + DocxCodeLang2MdStr[b.Code.Style.Language] + "\n")
 		buf.WriteString(strings.TrimSpace(p.ParseDocxBlockText(b.Code)))
 		buf.WriteString("\n```")
 	case lark.DocxBlockTypeQuote:
@@ -382,7 +452,7 @@ func (p *Parser) ParseDocxBlock(b *lark.DocxBlock, blockMap *orderedmap.OrderedM
 		buf.WriteString(p.ParseDocxBlockTableCell(b.BlockID, blockMap))
 	case lark.DocxBlockTypeTable:
 		buf.WriteString(p.ParseDocxBlockTable(b.ParentID, b.Table, blockMap))
-  case lark.DocxBlockTypeQuoteContainer:
+	case lark.DocxBlockTypeQuoteContainer:
 		buf.WriteString(p.ParseDocxBlockQuoteContainer(b.BlockID, b.QuoteContainer, blockMap))
 	default:
 		return ""
@@ -512,7 +582,7 @@ func (p *Parser) ParseDocxBlockTable(documentId string, t *lark.DocxBlockTable, 
 }
 
 func (p *Parser) ParseDocxBlockQuoteContainer(blockId string, q *lark.DocxBlocQuoteContainer, blockMap *orderedmap.OrderedMap) string {
-  contents := "> "
+	contents := "> "
 	for _, key := range blockMap.Keys() {
 		value, ok := blockMap.Get(key)
 		if !ok {

--- a/core/parser.go
+++ b/core/parser.go
@@ -370,7 +370,7 @@ func (p *Parser) ParseDocxBlock(b *lark.DocxBlock, blockMap *orderedmap.OrderedM
 		buf.WriteString(p.ParseDocxBlockText(b.Equation))
 		buf.WriteString("\n$$")
 	case lark.DocxBlockTypeTodo:
-		if b.Ordered.Style.Done {
+		if b.Todo.Style.Done {
 			buf.WriteString("- [x] ")
 		} else {
 			buf.WriteString("- [ ] ")

--- a/main.go
+++ b/main.go
@@ -133,9 +133,14 @@ func handleUrlArgument(url string, verbose bool) error {
 }
 
 func main() {
+	version, err := os.ReadFile("VERSION")
+	if err != nil {
+		version = []byte("v1-test")
+	}
+
 	app := &cli.App{
 		Name:    "feishu2md",
-		Version: "v1.3.4",
+		Version: strings.TrimSpace(string(version)),
 		Usage:   "download feishu/larksuite document to markdown file",
 		Flags: []cli.Flag{
 			&cli.BoolFlag{

